addons/web_editor/static/lib/odoo-editor/src/commands/commands.js

```diff
@@ -345,12 +347,12 @@ export const editorCommands = {
 
     // Change tags
     setTag(editor, tagName) {
-        const restoreCursor = preserveCursor(editor.document);
         const range = getDeepRange(editor.editable, { correctTripleClick: true });
         const selectedBlocks = [...new Set(getTraversedNodes(editor.editable, range).map(closestBlock))];
         const deepestSelectedBlocks = selectedBlocks.filter(block => (
             !descendants(block).some(descendant => selectedBlocks.includes(descendant))
         ));
+        const [anchorNode, anchorOffset, focusNode, focusOffset] = [firstLeaf(range.startContainer), range.startOffset, lastLeaf(range.endContainer), range.endOffset];
```

Comment:
@dmo-odoo 
Hello,
I had tried using getDeepRange first, but it was unable to maintain the cursor position when we change the element tags, such as from `<p>` to `<h1>,` while keeping our cursor in the middle of the text node. 
For example:

`<p>abc[]def</p> ---> <h1>abc[]def</h1>` (intended),

but by using `getDeepRange(editor.editable,{select: true})`, we will get something like this:

`<p>abc[]def</p> ---> <h1>[]abcdef</h1>`.

In this solution, we take the cursor offsets and then set them back to their previous position where they belong.

Reply:
Do you think you might be able to find the bug in `getDeepRange` that prevents it to work in this case and fix it ?

Reply:
@dmo-odoo 
Yes, I will definitely give it a try :relaxed:

Reply:
@dmo-odoo 
I looked into it and found that, `getDeepRange` was working as expected but the range that I was passing was incorrect, so I stored the range and passed the new and updated range and now it's working fine.

Reply:
@maad-odoo Thanks for investigating. However, what I'm trying to remove is the manual call to `firstLeaf` and `lastLeaf` since that is part of what `getDeepRange` should be doing. Could you try this please ?
```js
const range = getDeepRange(editor.editable, { select: true });
const restoreCursor = preserveCursor(editor.document);
// ... The for loop.
restoreCursor();
editor.historyStep();

Reply:
@dmo-odoo 
I tried what you suggested and it is not working because we are preserving the cursor position before the node is added (or updated) in the DOM. And when we restore the cursor position we get the old position.

Reply:
Weird... can you try by giving `getDeepRange` an addition option for `splitText: true` ?

Reply:
@dmo-odoo  
I tried getDeepRange with `splitText: true`, it is maintaining the cursor position, but we are not able to remove the nested blockquotes.

Reply:
Nested blockquotes ? What do you mean ?

Reply:
@dmo-odoo 
In this task, we should be able to remove the `blockquote` element that is nested inside another element when using the `deleteBackwards` operation.

For example:
If we insert a `blockquote` element inside a `<td>` and then hit `deleteBackward`, it does not remove that element.

The root cause of this issue is that when we insert the `blockquote`, the `setTag` method is called. In this method, the cursor position was previously preserved and then restored. The change that occurs in the DOM happens after preserving the cursor. As a result, when it is restored, our cursor is inside the `<td>` element and not inside the `blockquote` which will not let us delete the `blockquote`.

Reply:
@Zinston There is something fishy going on here because we can't seem to find a combination of `getDeepRange` and `preserveCursor` that works in this case. I feel like this might be related to the `preserveCursor` fixes that you have been working on recently. Could you give this bug a try with your fixed `preserveCursor` and see if there is a combination that works ?

Reply:
@dmo-odoo If you're referring to these fixes: https://github.com/odoo/odoo/pull/128562 they haven't been merged yet, the task has been greened in Testing India 30 days ago and waiting to be picked up by a tester in Belgium: https://www.odoo.com/web#id=3143042&cids=1&menu_id=4720&action=333&active_id=1695&model=project.task&view_type=form
You want me to cherrypick this onto my branch so I can see if it fixes it? If so, are these the steps I need to test: https://github.com/odoo/odoo/pull/123413#discussion_r1270246209 ?

Reply:
@Zinston Yes, I'm talking about that PR. However, it's not as easy as cherry-picking one commit, it's more like retrying to fix the bug using `getDeepRange` and `preserveCursor` and hope it works after the changes you've made to `prepareUpdate`. But now that I see your PR, I don't think it will solve this bug, actually, as we do not seem to use `prepareUpdate`. It is possible I had `prepareUpdate` and `preserveCursor` mixed up.

Anyway, I still feel like there is something fishy in how `prepareCursor` is behaving in this case. Would you mind having a look ? I feel like you might be able to uncover missing cases in `prepareCursor` like you did for `prepareUpdate`.

Reply:
@maad-odoo FYI I just r+'d [the other PR](https://github.com/odoo/odoo/pull/128562) which might have fixed the issue that you had with `preserveCursor`. You'll need to wait for it to be merged but afterwards you can rebase your branch and hopefully it will be fixed.

Reply:
@dmo-odoo I just rebased and checked into my branch, but the bug is still present.

Reply:
@maad-odoo Just to be sure we understand each other, I was not talking about the bug that this PR aims to solve. I was talking about the wrong behaviors of `prepareUpdate` that you had observed and that prevented you from using it to solve the bug.

Reply:
@dmo-odoo 
Yes, I did understand that you were not talking about the bug. Instead, I encountered an issue with preserveCursor not updating the cursor position once a blockQuote is inserted. The reason why I have not used preserveCursor in this fix is that we are preserving the cursor before the blockquote is inserted in the DOM. This results in us restoring our cursor position to the old cursor positions where blockquote was not present.

Reply:
@dmo-odoo A gentle reminder